### PR TITLE
Merge Contours

### PR DESCRIPTION
**Pull request by sbendjebla**
* **Please check if the PR fulfills these requirements**
- [ ] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Bug fix:
- [ ] New features:
- [ ] Documentation
- [ ] Optimization

* **Other information**:

_I improved the methods for **merging contours** that share primitives: the merge's result consider inner and outer contours.
I added **2 scripts** to illustrate the merge of contours, and all the related methods._ 
Merge_contours

### CODE REVIEW

File: volmdlr/wires.py [hunk 1]
```diff
@@ -1731,8 +1803,16 @@ def divide(self, contours, inside):
...
+    def merge_with(self, contour2d):
+        '''
+        merge two adjacent contours, sharing primitives, and returns one outer contour and inner contours (if there are any)
+        '''
+
+        merged_primitives = self.merge_primitives_with(contour2d)
+        contours = volmdlr.wires.Contour2D.contours_from_edges(merged_primitives)
+        contours = sorted(contours, key=lambda contour: contour.area(), reverse=True)
+
+        return contours
```
Comment (AxelRinghausen) [inline volmdlr/wires.py:1815]: Maybe Contour2D.merge_with could return a Surface2D ? What about Contour3D ?

**Comment by AxelRinghausen** [comment]
@sbendjebla I could really use a Contour2D.merge_with(contour**s**2d), contours2d being a list of Contour2D, and not just one single Contour2D. I don't know if it is easy to code, maybe we should discuss about it. What do you think ?

File: volmdlr/wires.py [hunk 2]
```diff
@@ -904,66 +903,67 @@ def shared_primitives_extremities(self, contour):
...
+                if edge1.point_belongs(edge2.start, 1e-5):
+                    # list_p.add(edge2.start)
+                    # instead of point not in list_p (due to errors)
+                    if list_p == []:
+                        list_p.append(edge2.start)
+                    if list_p != [] and edge2.start.point_distance(edge2.start.nearest_point(list_p)) > 1e-4:
+                        list_p.append(edge2.start)
+    
+                    # edges2.add(edge2)
+                    edges1.add(edge1)
```
Comment (AxelRinghausen) [inline volmdlr/wires.py:924]: If you're looking to refactor this, you could maybe iterate on the pairs of (edge, point), followed by a single if statement.

File: volmdlr/wires.py [hunk 3]
```diff
@@ -904,66 +903,67 @@ def shared_primitives_extremities(self, contour):
...
+            edges1 = list(edges1)
+            # edges2 = list(edges2)
 
-                edges1.add(edge1)
-                # edges2.add(edge2)
+            return edges1, list_p
 
-        edges1 = list(edges1)
-        # edges2 = list(edges2)
+        def order_extremities_points(self, contour):
```
Comment (AxelRinghausen) [inline volmdlr/wires.py:964]: Same goes for here.

**Review by AxelRinghausen — CHANGES_REQUESTED** [review]

File: volmdlr/wires.py [hunk 4]
```diff
@@ -904,66 +903,67 @@ def shared_primitives_extremities(self, contour):
...
-                edges1.add(edge1)
-
-            elif edge2.point_belongs(edge1.start):
-                # list_p.add(edge1.start)
-                # instead of point not in list_p (due to errors)
-                if list_p == []:
-                    list_p.append(edge1.start)
-                if list_p != [] and edge1.start.point_distance(edge1.start.nearest_point(list_p)) > 1e-4:
-                    list_p.append(edge1.start)
+        def possible_shared_primitives(self, contour):
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/wires.py]: I tried to refactor this method differently, but, it gives **C** as complexity index.

File: volmdlr/wires.py (hunk 3 above)
Comment (sbendjebla) [reply volmdlr/wires.py]: I tried to refactor this method differently, but, it gives **C** as complexity index.

**Review by masfaraud — APPROVED** [review]

**Review by masfaraud — APPROVED** [review]

**Pull request merged**